### PR TITLE
Pe 300 leaks

### PR DESCRIPTION
**Pull request by jbrezmorf**

### CODE REVIEW

File: src/flow/darcy_flow_mh.cc [hunk 1]
```diff
@@ -712,17 +719,14 @@ void DarcyMH::allocate_mh_matrix()
         ls->mat_set_values(loc_size, local_dofs, loc_size, local_dofs, zeros);
         
 
-        std::vector<int> tmp_rows;
-        tmp_rows.reserve(200);
-
         // compatible neighborings rows
         unsigned int n_neighs = ele_ac.full_iter()->n_neighs_vb;
-        for (unsigned int i = 0; i < n_neighs; i++) {
+        for (i = 0; i < n_neighs; i++) {
```
Comment (jbrezmorf) [inline src/flow/darcy_flow_mh.cc]: Why removed 'unsigned int'?

File: src/flow/darcy_flow_mh.cc [hunk 2]
```diff
@@ -712,17 +719,14 @@ void DarcyMH::allocate_mh_matrix()
...
             // every compatible connection adds a 2x2 matrix involving
             // current element pressure  and a connected edge pressure
             Neighbour *ngh = ele_ac.full_iter()->neigh_vb[i];
             int neigh_edge_row = mh_dh.row_4_edge[ ngh->edge_idx() ];
-            tmp_rows.push_back(neigh_edge_row);
+            tmp_rows[i] = neigh_edge_row;
```
Comment (jbrezmorf) [inline src/flow/darcy_flow_mh.cc]: This is wrong, reserve is not resize. tmp_rows have still size 0.

File: src/flow/darcy_flow_assembly.hh [hunk 3]
```diff
@@ -150,7 +150,9 @@ public:
         assemble_element(ele_ac);
         assemble_source_term(ele_ac);
         
-        
+        //due to time dependent flow, fill the whole diagonal (otherwise it is allocated)
```
Comment (jbrezmorf) [inline src/flow/darcy_flow_assembly.hh]: So it is necessary to keep matrix structure unchanged? But what can change for unsteady case? Only possibility is switching from steady to unsteady. Comment seems misleading since this have noting to do with allocation, but with allowing reuse of matrix structure.

File: src/la/local_system.hh [hunk 4]
```diff
@@ -28,6 +28,8 @@ class LocalSystem
 public:
     typedef arma::uvec DofVec;
     
+    static constexpr double artificial_zero = std::numeric_limits<double>::min();
```
Comment (jbrezmorf) [inline src/la/local_system.hh]: What about 'almost_zero', thats more descriptive. Also add detailed explanation of the whole problem here together with suggestions where to use it.

File: unit_tests/la/schur_compl_test.cpp [hunk 5]
```diff
@@ -135,7 +135,7 @@ TEST(schur, complement) {
 		const PetscScalar *vals;
 		for (unsigned int i=0; i<block_size; i++) {
 			MatGetRow(schurComplement->get_a_inv(), i + rank*block_size, &ncols, &cols, &vals);
-			EXPECT_FLOAT_EQ( (1.0 / (double)(rank + 2)), vals[i] );
+			EXPECT_FLOAT_EQ( (1.0 / (double)(rank + 2)), vals[0] );
```
Comment (jbrezmorf) [inline unit_tests/la/schur_compl_test.cpp]: Why this change?

Comment (Paulie14) [reply unit_tests/la/schur_compl_test.cpp]: Thanks to setting `MatSetOption(matrix_, MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE);`,
the matrix `get_a_inv()` is now sparser:
``` 
row 0: (0, 0.5) 
row 1: (1, 0.5)
```
therefore vector vals has NOT size==block_size, but only 1 (only the diagonal value is nonzero in the test)

File: src/flow/darcy_flow_assembly.hh (hunk 3 above)
Comment (Paulie14) [reply src/flow/darcy_flow_assembly.hh]: This was a fast fix:
- in steady case, we want the zeros on the diagonal to be ignored (and not allocated)
- in unsteady case, if the zeros on the diagonal are ignored (and not allocated), then it crashes when assembling time terms (in `setup_time_term()`) because we need to write to the diagonal and not enough memory is allocated in that rows..

Any better suggestions, what to do with it?

File: src/flow/darcy_flow_mh.cc (hunk 2 above)
Comment (Paulie14) [reply src/flow/darcy_flow_mh.cc]: agree, my bad..

**Comment by Paulie14** [comment]
Here, we are solving issue #875.

**Comment by Paulie14** [comment]
Please see the last commit 8f9b5a9 for the sparsity pattern.

**Comment by Paulie14** [comment]
Ready to merge now.
